### ISSUE TITLE
Windows: remove direct HCS calls

### ISSUE THREAD

**Issue by cpuguy83** [moby/moby#41455]
Currently dockerd maks calls directly to hcs (v1) rather than using containerd for managing containers.

containerd now has support for Windows through a windows hcsshim for containerd. This also happens to use hcs v2 which fixes a lot of issues/shortcomings of hcs v1.

Let's work towards moving those direct HCS calls to containerd calls. This should eliminate some branching in the container management code.

**Comment by olljanat** [comment]
Looks to be one more good reason to fix those Windows integration tests before that work starts...

**Comment by thaJeztah** [comment]
We do have experimental support for containerd, correct? (but don't ship containerd yet for windows)

**Comment by cpuguy83** [comment]
There is some stuff... really old and untested stuff.

**Comment by olljanat** [comment]
Yes there is and PR also contains also some info how to test https://github.com/moby/moby/pull/38541

Btw. Does switch to hcs v2 also mean end of Windows Server 2016 support? Based on [this](https://support.microsoft.com/en-us/lifecycle/search/19649) its EOS is on January 2022 so we might need some depreciation Notice?

**Comment by cpuguy83** [comment]
@olljanat I'm not sure Docker should be tied to Windows support lifecycles.

That said it is possible we could keep hcsv1 calls behind a `--runtime` configuration just like we do with linux and shimv1, and indeed this might be a good way to transition and provide a way to bail out in case of problems with the containerd hcs shim.

**Comment by olljanat** [comment]
@cpuguy83 probably not. Just thinking that how to make things clear for users but maybe we can include check to code that RS5 is minimum requirement on 21.x.

What comes to multiple runtimes I probably would prefer to keep things simple switch Windows (and maybe even Linux) use containerd only on 21.x and just extend 20.x lifecycle a bit longer if needed.

**Comment by olljanat** [comment]
Just to check that I understand this right, libnetwork will still need call HCS directly or will that logic also be on containerd (or some other place)?

**Comment by cpuguy83** [comment]
No networking in containerd. So we'll still need to do whatever we do with libnetwork and hcs.

**Comment by cpuguy83** [comment]
Started messing around with this in (WIP) #42089 

Previously @lowenna started integrating containerd for the Windows side, so there is *some* existing code for this, I just don't know the state of it yet.
And I'm sure CI is lacking containerd bins to test this right now.

**Comment by olljanat** [comment]
#41479 contains CI draft for this. It would be good if someone who have access to build servers would investigate that where it fails.

**Comment by TBBle** [comment]
Given that Docker Desktop 3.2.0 (with Docker Engine 20.10.3) just dropped support for Windows 1709, there's only one currently-supported-by-Docker Windows desktop version that doesn't support HCS v2, that'll be Windows 10 1803, and even the Enterprise builds of that [end support in May 2020](https://docs.microsoft.com/en-us/windows/release-health/release-information).

**Comment by olljanat** [comment]
> @olljanat I'm not sure Docker should be tied to Windows support lifecycles.
> 
> That said it is possible we could keep hcsv1 calls behind a `--runtime` configuration just like we do with linux and shimv1, and indeed this might be a good way to transition and provide a way to bail out in case of problems with the containerd hcs shim.

Btw. Maybe the fact that new Windows Server version is coming will help us on here. What if we just decide that >= Win Srv 2022 (and matching Win 10 version) will use containerd default and leave older versions to use HCS v1? And then we can also add logic which prevents that old LCOW to be enabled with those >= 2022 versions.

That would simplify testing and recude risk level on change.

**Comment by weijuans-msft** [comment]
@olljanat great idea on leveraging the new Windows Server 2022 release motion. It may hurt a bit :), but we can work together to ensure folks getting the memo and not panic. And for other in-market supported releases, they can stay on docker/HCS v1 as default, but we offer the run-time switching options in #42089 and maybe also change it into using HCSv2/containerd as default later on.

**Comment by olljanat** [comment]
@weijuans-msft #41479 now contains CI support for Windows + ContainerD combination and defaults to ContainerD on Windows Server 2022 but based on latest comments https://github.com/moby/moby/pull/41479#issuecomment-843635491 it looks to be still unclear if that default changes are now wanted or not. Also here looks to be some issue where connection between Docker and ContainerD gets stuck on certain cases https://github.com/moby/moby/pull/41479#discussion_r636967589

Can you find someone who is familiar with ContainerD's Windows implementation and help get those problematic parts fixed?

**Comment by TBBle** [comment]
For some reason I can't reply directly to https://github.com/moby/moby/pull/41479#discussion_r584795703, but AFAIK that stdin hang wasn't resolved, and I suspect I'm now seeing the same or a related issue directly in containerd: https://github.com/containerd/containerd/issues/6652
